### PR TITLE
build(fpb-lint): linting errors as PR comments, resolved #4416

### LINKED ISSUE #4416: Linting Errors as PR Comments
I think it'd be worth pulling the lint summary from CI and posting them as a comment in the PR?

Many contributors either don't know how or don't bother to check the failed checks. And some that do, don't understand what's wrong because they're only reading the bottom line, or just don't know what to focus on.

It'd massively reduce the workload of maintainers to push the warnings to a PR comment, so the relevant information that contributors need is readily available.

### PR DESCRIPTION
**Pull request by LuigiImVector**
## What does this PR do?
Improve repo

### Description

I improved the fpb-lint.yml workflow so that linter errors were shown via comment, as proposed by @SethFalco on #4416.

In this PR LuigiImVector#10 I conducted some tests, all successful.

## Checklist:
- [X] Read our [contributing guidelines](https://github.com/EbookFoundation/free-programming-books/blob/main/docs/CONTRIBUTING.md)

## Follow-up

- Check the status of GitHub Actions and resolve any reported warnings!

### CODE REVIEW

File: .github/workflows/fpb-lint.yml
Comment (davorpa) [inline .github/workflows/fpb-lint.yml:20]: Not needed since uses `actions/github-script` to clean output

Comment (LuigiImVector) [reply .github/workflows/fpb-lint.yml:20]: oh right, I forgot to remove it, thanks

File: .github/workflows/fpb-lint.yml
Comment (davorpa) [inline .github/workflows/fpb-lint.yml:40]: Is this line to create a empty file that can be used at line 60?, Better use `touch` I tought

```shell
touch error.log
```

Comment (LuigiImVector) [reply .github/workflows/fpb-lint.yml:40]: You're right, create a code suggestion

File: .github/workflows/fpb-lint.yml
Comment (davorpa) [inline .github/workflows/fpb-lint.yml:11]: About current workflow error at `Print output` task.... 

image

Reading https://www.damirscorner.com/blog/posts/20220506-TokenPermissionsForGithubActions.html... a bit I think that is about GitHub token permissions

See Seth workarround at PR #5557 : https://github.com/SethFalco/free-programming-books/blob/e9f7dff3be87e640a72f97bd4eec1be79ec2b43a/.github/workflows/fpb-lint.yml#L7-L9

**Comment by SethFalco** [comment]
Might be worth evaluating this one more.

The problem I ran into with my one which through me off a bit was that if you look at your tests carefully, you executed all of those comments from a PR.

A PR that could just as easily modify the workflow file to do whatever they want as part of their PR, include spam or execute other actions if given write access to the repo.

Before we merge this, would you mind merging the one in your fork and perhaps I can try do a malicious PR to it and we'll see what happens? 🤔

File: .github/workflows/fpb-lint.yml
Comment (SethFalco) [reply .github/workflows/fpb-lint.yml:11]: Note, my workaround is insecure and shouldn't be used if I understand correctly.

The real solution is in my comment below. (Split into 2 separate workflows, one creates the message, the other reads the message and posts the review.)

**Comment by LuigiImVector** [comment]
> Before we merge this, would you mind merging the one in your fork, and perhaps I can try to do a malicious PR to it, and we'll see what happens? 🤔

Now I'm going to read the article you linked, in the meantime is it enough for me to write?
`on: [push, pull_request_target]`

I'm not sure if I understand correctly

**Comment by SethFalco** [comment]
I recommend you scroll down to where it says:

> Below is an example of the intended usage in which the results of an unprivileged pull_request workflow are combined with a privileged workflow to leave a comment in response to a received PR:

Over here you'll see an example of it using 2 workflows to be able to leave comments/reviews without creating the security issue.

**Comment by LuigiImVector** [comment]
> Over here you'll see an example of it using 2 workflows to be able to leave comments/reviews without creating the security issue.

_"Incoming data from artifacts is potentially untrusted. When used in a safe manner, like reading PR numbers or reading a code coverage text to comment on the PR, it is safe to use such untrusted data in the privileged workflow context. However if the artifacts were, for example, binaries built from an untrusted PR, it would be a security vulnerability to run them in the privileged workflow_run workflow context. Artifacts resulting from untrusted PR data are themselves untrusted and should be treated as such when handled in privileged contexts."_ 

Even this could not be a totally safe solution

---

Maybe it's enougth to [disable the auto-run](https://docs.github.com/en/organizations/managing-organization-settings/disabling-or-limiting-github-actions-for-your-organization#configuring-required-approval-for-workflows-from-public-forks) of the workflow, giving us the ability to approve only non-suspicious PR

**Comment by SethFalco** [comment]
> When used in a safe manner, like reading PR numbers or reading a code coverage text to comment on the PR, it is safe to use such untrusted data in the privileged workflow context

> However if the artifacts were, for example, binaries built from an untrusted PR, it would be a security vulnerability to run them in the privileged workflow_run workflow context

You're right, but in this case, it's fine because we're just reading a text file, not executing a binary. Good catch though, it's definitely good to be vigilant just to make sure we're 100% sure we're not screwing this repo up. XD

**Comment by LuigiImVector** [comment]
So I'll try to implement the 2 workflows here

**Comment by LuigiImVector** [comment]
I divided the workflow and now the `free-programming-books-lint` doesn't any return error; `workflow_run` is only triggered in the default branch

File: .github/workflows/comment-pr.yml
Comment (SethFalco) [inline .github/workflows/comment-pr.yml:13]:
```suggestion
  upload:
    permissions:
      pull-requests: write
```

This job would need to explicitly declare write access for `pull-request` here, I think?
Either that, or in the repo settings someone would have to enable write access by default for all workflows, but that's probably not a good idea for a public repo.

File: .github/workflows/fpb-lint.yml
Comment (SethFalco) [inline .github/workflows/fpb-lint.yml:35]:
```suggestion
```

If my other suggestions are merged, these line can be deleted.

**Review by SethFalco — COMMENTED** [review]
I've tested all the suggestions in this repo:
https://github.com/SethFalco/free-programming-books/pull/6

---

There is another issue. Since we have 2 workflows named `free-programming-books-lint`, one for `push` and one for `pull-request`, the comment action gets triggered twice, with one always failing since the artifacts don't exist.

I don't think we really need this on `push` anyway, so perhaps we can drop that one and make this run on `pull-request` only?

---

Giving it more thought, I originally proposed that we should have the action approve or request. However, I think that might be a bad idea, and we should instead only post a comment.

Since we have many other checks in this repository in other actions, it'd look odd if this action approved the pull request, while other linter actions failed. Contributors will receive contradicting information and won't be sure whether they need to fix something or not. We should probably just post a comment, and later we can try to hide it once it's outdated.

This can be done with the following change:

```diff
-             gh pr review $(<PRurl) -r -b "Linter failed, fix the error(s):
+             gh pr comment $(<PRurl) -b "Linter failed, fix the error(s):
              \`\`\`
              $(cat error.log)
              \`\`\`"
              gh pr edit $(<PRurl) --add-label "linter error"
            else
-             gh pr review $(<PRurl) -a
```

**Review by Milad335t — APPROVED** [review]

**Review by Milad335t — APPROVED** [review]

**Pull request closed without merging**
